WASI support

tessi opened the pull request:
This PR adds support to run WASM modules compiled to WASI. We will support passing program arguments and environment variables to WASI programs. Support for preopened files and mapped directories is not planned in this PR (and may be added later).

When given a `wasi` map during instance initialization, we now fill the `wasi_snapshot_preview1` namespace with [`wasmer-wasi` default implementations](https://github.com/wasmerio/wasmer/blob/8c174554568c716fe3d0530a7ebb2fb4f9f4b023/lib/wasi/src/lib.rs#L230). These defaults implement [the latest WASI spec](https://github.com/WebAssembly/WASI/blob/master/phases/snapshot/docs.md).

It is possible to overwrite the default implementations in the `wasi_snapshot_preview1` with callbacks implemented in Elixir.

tessi commented on lib/wasmex/instance.ex:
```diff
@@ -49,6 +49,16 @@ defmodule Wasmex.Instance do
+  @spec wasi_from_bytes(binary(), %{optional(binary()) => (... -> any())}, %{optional(:args) => [String.t()], optional(:env) => %{String.t() => String.t()}}) ::
+          {:error, binary()} | {:ok, __MODULE__.t()}
+  def wasi_from_bytes(bytes, imports, wasi) when is_binary(bytes) and is_map(imports) and is_map(wasi) do
+    
```
todo: here I need to take apart the `wasi` map:

* args: need to make sure its a list of strings
* env: need to map the `Map` from `name => value` pairs to a list of `name=value` strings and make sure neither `name` nor `value` include an equals sign (`=`)

tessi commented on lib/wasmex/instance.ex:
```diff
@@ -49,6 +49,16 @@ defmodule Wasmex.Instance do
+          {:error, binary()} | {:ok, __MODULE__.t()}
+  def wasi_from_bytes(bytes, imports, wasi) when is_binary(bytes) and is_map(imports) and is_map(wasi) do
+    
+    case Wasmex.Native.instance_new_from_bytes(bytes, imports, args, env) do
```
```suggestion
    case Wasmex.Native.instance_new_wasi_from_bytes(bytes, imports, args, env) do
```

myobie commented on native/wasmex/src/instance.rs:
```diff
@@ -22,14 +24,48 @@ pub struct InstanceResource {
 // * bytes (binary): the bytes of the WASM module
 // * imports (map): a map defining eventual instance imports, may be empty if there are none.
-//   structure: %{namespace_name: %{import_name: {TODO: signature}}}
+//   structure: %{namespace_name: %{import_name: {:fn, param_types, result_types, captured_function}}}
```
Oh OK, ✅ marked as done 😎

tessi replied:
I didn't check this against reality yet (wrote it from memory when I drafted the PR).

So making sure this is actually correct is still a Todo :)

tessi commented:
I added a sample WASM/WASI project in `test/wasi_test` which is now compiled to wasm/wasi on every test run.

It prints the current time and all arguments and environment variables:

```rust
use std::env;
use std::time::{SystemTime, UNIX_EPOCH};

fn main() {
    println!("Hello from the WASI test program!");
    println!();

    println!("Arguments:");
    for arg in env::args().collect::<Vec<String>>() {
        println!("{}", arg);
    }
    println!();

    println!("Environment:");
    for (name, value) in env::vars().collect::<Vec<(String, String)>>() {
        println!("{}={}", name, value);
    }
    println!();

    println!("Current Time (Since Unix Epoch):");
    let seconds_since_epoch = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .expect("Time went backwards!")
        .as_secs();
    println!("{}", seconds_since_epoch);
}
```

calling this program from elixir is possible as done in our integration test:

```elixir
test "runnung a WASM/WASI module" do
    imports = %{}

    wasi = %{
      args: ["hello", "from elixir"],
      env: %{
        "A_NAME_MAPS" => "TO_A_VALUE",
        "THE_TEST_WASI_FILE" => "PRINTS_ALL_ENVS"
      }
    }

    instance = start_supervised!({Wasmex, %{bytes: @wasi_test_bytes, imports: imports, wasi: wasi}})
    {:ok, _} = Wasmex.call_function(instance, :_start, [])
  end
```

```
❯ mix test test/wasmex_test.exs:244
Compiling NIF crate :wasmex (native/wasmex)...
    Finished dev [unoptimized + debuginfo] target(s) in 0.12s
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
Excluding tags: [:test]
Including tags: [line: "244"]

Hello from the WASI test program!

Arguments:
hello
from elixir

Environment:
A_NAME_MAPS=TO_A_VALUE
THE_TEST_WASI_FILE=PRINTS_ALL_ENVS

Current Time (Since Unix Epoch):
1588773509
.

Finished in 6.3 seconds
20 tests, 0 failures, 19 excluded
```

tessi commented:
What is missing at this point is:

* documentation and more tests
* overwriting some WASI functions with elixir implementations

I decided to limit the scope of this PR to only support args and env variables (no file descriptors or directories).

tessi commented:
Overwriting default WASI functions with an elixir callback currently fails because of an [issue in wasmer](https://github.com/wasmerio/wasmer/issues/1430). I try to resolve it there in parallel to this PR.

I think we have to wait with this PR until the wasmer bug is resolve because I think we need these overwrites to test our WASI support (we need to overwrite `fd_write` to capture stdout of the WASM instance).

tessi commented:
Update: My rust-foo is not strong enough to fix the wasmer-bug myself. I got no response yet on my question how the wasmer team prioritized the soundness error. My best guess is that it will take a while.

tessi commented:
Update: With wasmer 1.0 the aforementioned bug is fixed. I had to refactor some code to cope with the new API, but things seem to work now. What is missing at the moment:

1. some code cleanup
2. polishing the elixir-facing API, so it feels easy to use/understand
3. finishing integration tests (finishing the integration test to capture stdout and compare it against an expected value) and adding new unit tests
4. think about other WASI API's (like mapping files and directories from Elixir, including stdout/in/err)

tessi commented:
Alright, after some time waiting I have another update.

Last time I struggled most with having a proper integration test for wasi modules (i didn't want to accept shipping code that works, but isn't tested). This is now resolved 🎉 

I introduces the concept of a `Pipe` which allows us to mock stdin/stdout/stderr of a WASI module (and potentially other files in the future). Using that, I could capture stdout of a test WASI module in our integration tests and do some assertions on it.

What is missing (for another night hacking on this PR on my side) is some code cleanups (DRY'ing up the rust side) and consistency/usability checks. I'd be very interested in your feedback if you'd like to try this out.

IMO, we could ship WASI support as an experimental feature (however we want to label that 🤔 ) until we figured out how to do file mapping and other more advanced WASI things. This would enable us to merge this PR before sneaking even more features in 🚢

tessi commented:
Alright, a few more updates went into this branch and I think it's ready to be merged.

* updated docs (incl. a shorter README)
* documented all WASI features
* added support for preopened files and incl. read/write/create permissions and aliases
* (integration-)tested this feature a little more
* some code-cleanup
